shared/src/sources/balancer/swap/fixed_point.rs

```diff
@@ -96,6 +96,10 @@ impl Debug for Bfp {
 }
 
 impl Bfp {
+    pub fn to_f64_lossy(self) -> f64 {
+        self.as_uint256().to_f64_lossy() / 1e18
+    }
```

Comment:
Not sure if a problem in practice but want to note that this introduces a lot of unnecessary floating point error because it first creates a large U256, converts it to float which is lossy and then divides by a large float which is lossy again.
A better way to do it is to get the integer amount first by dividing out 1e18 as U256 and then converting to float because that number will be much closer to 0. Then convert the rest (amount after comma) to U256 and only divide that by 1e18 in float.
Haven't fully thought this true so the right solution might be a little different but the concept should be the same.

Reply:
This is a great suggestion, I will do this. However, I would like to point out that this is only used inside the test cases introduced here. I should probably add `#[cfg(test)]` as a header to this method.

Reply:
Since its only used in tests, I would prefer adding `#[cfg(test)]` to avoid unnecessary for testing.

Reply:
It is 1 line versus 5 lines I think so not much extra complexity and could be useful for tests too to have lower error. But yeah no strong opinion as long as its used only in tests.

Reply:
So I tried out you suggestion here and found that the two results are identical. Please let me know if this is the approach you envisioned:

```rust
        let as_uint = self.as_uint256();
        let int_part = as_uint / *ONE_18;
        let decimal_part = as_uint - (int_part * *ONE_18);
        println!("Int {}, Dec {}", int_part, decimal_part);
        println!("OLD SOLUTION: {}", self.as_uint256().to_f64_lossy() / 1e18);
        println!("NEW SOLUTION: {}", int_part.to_f64_lossy() + decimal_part.to_f64_lossy() / 1e18);
```

Reply:
This is what I had in mind. It thought it would be equivalent for small numbers but I expected there to be differences for large ones.

Reply:
Which of these functions would be public? If it isn't all then it would also be fine to only check bounds in those higher level functions while the others have the invariant that all indexes are valid.